File: src/tracking.ts
import type { RelationDetection, SceneRelation } from "./relations";

export type TrackableDetection = RelationDetection & {
  group?: string;
};

export type TrackedDetection<T extends TrackableDetection = TrackableDetection> = T & {
  trackId: number;
  trackAge: number;
};

type InternalTrack<T extends TrackableDetection> = {
  id: number;
  detection: T;
  age: number;
  misses: number;
};

export type TrackerOptions = {
  iouThreshold?: number;
  maxMisses?: number;
};

function area(box: TrackableDetection) {
  return Math.max(0, box.x2 - box.x1) * Math.max(0, box.y2 - box.y1);
}

export function intersectionOverUnion(a: TrackableDetection, b: TrackableDetection) {
  const x1 = Math.max(a.x1, b.x1);
  const y1 = Math.max(a.y1, b.y1);
  const x2 = Math.min(a.x2, b.x2);
  const y2 = Math.min(a.y2, b.y2);
  const intersection = Math.max(0, x2 - x1) * Math.max(0, y2 - y1);
  if (!intersection) return 0;
  const union = area(a) + area(b) - intersection;
  return union > 0 ? intersection / union : 0;
}

export class IoUTracker<T extends TrackableDetection> {
  private readonly iouThreshold: number;
  private readonly maxMisses: number;
  private tracks: InternalTrack<T>[] = [];
  private nextId = 1;

  constructor(options: TrackerOptions = {}) {
    this.iouThreshold = options.iouThreshold ?? 0.28;
    this.maxMisses = options.maxMisses ?? 8;
  }

  reset() {
    this.tracks = [];
    this.nextId = 1;
  }

  update(detections: T[]): TrackedDetection<T>[] {
    const unmatchedTracks = new Set(this.tracks.map((_, index) => index));
    const assignments = new Map<number, number>();

    // Match high-confidence detections first. Label agreement plus IoU keeps
    // IDs stable enough for a lightweight browser tracker without adding a
    // second neural network.
    const detectionOrder = detections
      .map((detection, index) => ({ detection, index }))
      .sort((left, right) => right.detection.confidence - left.detection.confidence);

    for (const { detection, index: detectionIndex } of detectionOrder) {
      let bestTrackIndex = -1;
      let bestIou = this.iouThreshold;

      unmatchedTracks.forEach((trackIndex) => {
        const track = this.tracks[trackIndex];
        if (track.detection.label !== detection.label) return;
        const iou = intersectionOverUnion(track.detection, detection);
        if (iou > bestIou) {
          bestIou = iou;
          bestTrackIndex = trackIndex;
        }
      });

      if (bestTrackIndex >= 0) {
        assignments.set(detectionIndex, bestTrackIndex);
        unmatchedTracks.delete(bestTrackIndex);
      }
    }

    unmatchedTracks.forEach((trackIndex) => {
      this.tracks[trackIndex].misses += 1;
    });

    const output = detections.map((detection, detectionIndex) => {
      const matchedTrackIndex = assignments.get(detectionIndex);
      if (matchedTrackIndex !== undefined) {
        const track = this.tracks[matchedTrackIndex];
        track.detection = detection;
        track.age += 1;
        track.misses = 0;
        return { ...detection, trackId: track.id, trackAge: track.age };
      }

      const track: InternalTrack<T> = {
        id: this.nextId,
        detection,
        age: 1,
        misses: 0,
      };
      this.nextId += 1;
      this.tracks.push(track);
      return { ...detection, trackId: track.id, trackAge: track.age };
    });

    this.tracks = this.tracks.filter((track) => track.misses <= this.maxMisses);
    return output;
  }

  get activeTrackCount() {
    return this.tracks.filter((track) => track.misses === 0).length;
  }
}

type SmoothedEntry = {
  relation: SceneRelation;
  score: number;
  rankScore: number;
  hits: number;
  misses: number;
};

export type RelationSmootherOptions = {
  emaWeight?: number;
  maxMisses?: number;
  maxRelations?: number;
};

function relationKey(relation: SceneRelation) {
  const subjectId = relation.subject.trackId;
  const objectId = relation.object.trackId;
  if (subjectId === undefined || objectId === undefined) return null;
  return `${subjectId}|${relation.predicate}|${objectId}`;
}

export class RelationSmoother {
  private readonly emaWeight: number;
  private readonly maxMisses: number;
  private readonly maxRelations: number;
  private entries = new Map<string, SmoothedEntry>();

  constructor(options: RelationSmootherOptions = {}) {
    this.emaWeight = options.emaWeight ?? 0.62;
    this.maxMisses = options.maxMisses ?? 1;
    this.maxRelations = options.maxRelations ?? 8;
  }

  reset() {
    this.entries.clear();
  }

  update(relations: SceneRelation[]) {
    const seen = new Set<string>();

    for (const relation of relations) {
      const key = relationKey(relation);
      if (!key) continue;
      seen.add(key);
      const previous = this.entries.get(key);
      if (!previous) {
        this.entries.set(key, {
          relation,
          score: relation.score,
          rankScore: relation.rankScore,
          hits: 1,
          misses: 0,
        });
        continue;
      }

      const score = previous.score * this.emaWeight + relation.score * (1 - this.emaWeight);
      const rankScore = previous.rankScore * this.emaWeight + relation.rankScore * (1 - this.emaWeight);
      this.entries.set(key, {
        relation: { ...relation, score, rankScore },
        score,
        rankScore,
        hits: previous.hits + 1,
        misses: 0,
      });
    }

    for (const [key, entry] of this.entries) {
      if (seen.has(key)) continue;
      entry.misses += 1;
      if (entry.misses > this.maxMisses) this.entries.delete(key);
    }

    return this.values();
  }

  values() {
    return [...this.entries.values()]
      .sort((left, right) => right.rankScore - left.rankScore)
      .slice(0, this.maxRelations)
      .map((entry) => ({
        ...entry.relation,
        score: entry.score,
        rankScore: entry.rankScore,
      }));
  }
}

export function resolveRelationBoxes<T extends TrackedDetection>(
  relations: SceneRelation[],
  detections: T[],
) {
  const byId = new Map(detections.map((detection) => [detection.trackId, detection]));
  return relations
    .map((relation) => {
      const subjectId = relation.subject.trackId;
      const objectId = relation.object.trackId;
      if (subjectId === undefined || objectId === undefined) return relation;
      const subject = byId.get(subjectId);
      const object = byId.get(objectId);
      if (!subject || !object) return null;
      return {
        ...relation,
        subject: { ...relation.subject, ...subject },
        object: { ...relation.object, ...object },
      };
    })
    .filter((relation): relation is SceneRelation => relation !== null);
}

File: android/app/src/main/java/com/sunnygumber/cctvaivisionlab/tracking/Tracking.kt
package com.sunnygumber.cctvaivisionlab.tracking

import com.sunnygumber.cctvaivisionlab.core.BoundingBox
import com.sunnygumber.cctvaivisionlab.core.Detection
import com.sunnygumber.cctvaivisionlab.core.SceneRelation
import com.sunnygumber.cctvaivisionlab.core.Tracker
import kotlin.math.hypot
import kotlin.math.max
import kotlin.math.min

fun intersectionOverUnion(a: BoundingBox, b: BoundingBox): Float {
    val left = max(a.left, b.left)
    val top = max(a.top, b.top)
    val right = min(a.right, b.right)
    val bottom = min(a.bottom, b.bottom)
    val intersection = max(0f, right - left) * max(0f, bottom - top)
    if (intersection <= 0f) return 0f
    val union = a.area + b.area - intersection
    return if (union > 0f) intersection / union else 0f
}

private fun overlapOverSmallerArea(a: BoundingBox, b: BoundingBox): Float {
    val left = max(a.left, b.left)
    val top = max(a.top, b.top)
    val right = min(a.right, b.right)
    val bottom = min(a.bottom, b.bottom)
    val intersection = max(0f, right - left) * max(0f, bottom - top)
    if (intersection <= 0f) return 0f
    val smaller = min(a.area, b.area)
    return if (smaller > 0f) intersection / smaller else 0f
}

fun deduplicateDetections(
    detections: List<Detection>,
    iouThreshold: Float = 0.45f,
    containmentThreshold: Float = 0.72f,
): List<Detection> {
    val kept = ArrayList<Detection>()
    detections.sortedByDescending { it.confidence }.forEach { candidate ->
        val duplicate = kept.any { existing ->
            existing.classId == candidate.classId &&
                (intersectionOverUnion(existing.box, candidate.box) >= iouThreshold ||
                    overlapOverSmallerArea(existing.box, candidate.box) >= containmentThreshold)
        }
        if (!duplicate) kept += candidate
    }
    return kept
}

private data class InternalTrack(
    val id: Int,
    var detection: Detection,
    var age: Int,
    var misses: Int,
    var velocityX: Float = 0f,
    var velocityY: Float = 0f,
)

class IoUTracker(
    private val minimumIou: Float = 0.10f,
    private val minimumMatchScore: Float = 0.34f,
    private val maxMisses: Int = 6,
) : Tracker {
    private val tracks = ArrayList<InternalTrack>()
    private var nextId = 1

    override fun update(detections: List<Detection>): List<Detection> {
        val unmatchedTracks = tracks.indices.toMutableSet()
        val assignments = mutableMapOf<Int, Int>()

        detections.indices
            .sortedByDescending { detections[it].confidence }
            .forEach { detectionIndex ->
                val detection = detections[detectionIndex]
                var bestTrack = -1
                var bestScore = minimumMatchScore

                unmatchedTracks.forEach { trackIndex ->
                    val track = tracks[trackIndex]
                    if (track.detection.classId != detection.classId) return@forEach

                    val score = matchScore(track, detection)
                    if (score > bestScore) {
                        bestScore = score
                        bestTrack = trackIndex
                    }
                }

                if (bestTrack >= 0) {
                    assignments[detectionIndex] = bestTrack
                    unmatchedTracks.remove(bestTrack)
                }
            }

        unmatchedTracks.forEach { tracks[it].misses += 1 }

        val output = detections.mapIndexed { index, rawDetection ->
            val matched = assignments[index]
            if (matched != null) {
                val track = tracks[matched]
                val oldBox = track.detection.box
                val dx = rawDetection.box.centerX - oldBox.centerX
                val dy = rawDetection.box.centerY - oldBox.centerY

                track.velocityX = track.velocityX * 0.55f + dx * 0.45f
                track.velocityY = track.velocityY * 0.55f + dy * 0.45f

                val stabilized = rawDetection.copy(
                    box = blendBoxes(oldBox, rawDetection.box, newWeight = 0.72f),
                    trackId = track.id,
                    trackAge = track.age + 1,
                )
                track.detection = stabilized
                track.age += 1
                track.misses = 0
                stabilized
            } else {
                val tracked = rawDetection.copy(trackId = nextId, trackAge = 1)
                tracks += InternalTrack(
                    id = nextId,
                    detection = tracked,
                    age = 1,
                    misses = 0,
                )
                nextId += 1
                tracked
            }
        }

        tracks.removeAll { it.misses > maxMisses }
        return output
    }

    private fun matchScore(track: InternalTrack, detection: Detection): Float {
        val predicted = shiftBox(
            track.detection.box,
            track.velocityX * (track.misses + 1),
            track.velocityY * (track.misses + 1),
        )
        val candidate = detection.box
        val iou = intersectionOverUnion(predicted, candidate)

        val distance = hypot(
            (predicted.centerX - candidate.centerX).toDouble(),
            (predicted.centerY - candidate.centerY).toDouble(),
        ).toFloat()
        val scale = max(
            max(predicted.width, candidate.width),
            max(predicted.height, candidate.height),
        ).coerceAtLeast(1f)
        val centerScore = (1f - distance / (scale * 1.65f)).coerceIn(0f, 1f)

        val largerArea = max(predicted.area, candidate.area).coerceAtLeast(1f)
        val sizeScore = (min(predicted.area, candidate.area) / largerArea).coerceIn(0f, 1f)

        if (iou < minimumIou && centerScore < 0.44f) return Float.NEGATIVE_INFINITY
        if (sizeScore < 0.28f) return Float.NEGATIVE_INFINITY

        return iou * 0.58f +
            centerScore * 0.30f +
            sizeScore * 0.12f -
            track.misses * 0.025f
    }

    override fun reset() {
        tracks.clear()
        nextId = 1
    }

    val activeTrackCount: Int
        get() = tracks.count { it.misses == 0 }

    private fun shiftBox(box: BoundingBox, dx: Float, dy: Float) = BoundingBox(
        left = box.left + dx,
        top = box.top + dy,
        right = box.right + dx,
        bottom = box.bottom + dy,
    )

    private fun blendBoxes(old: BoundingBox, new: BoundingBox, newWeight: Float): BoundingBox {
        val oldWeight = 1f - newWeight
        return BoundingBox(
            left = old.left * oldWeight + new.left * newWeight,
            top = old.top * oldWeight + new.top * newWeight,
            right = old.right * oldWeight + new.right * newWeight,
            bottom = old.bottom * oldWeight + new.bottom * newWeight,
        )
    }
}

class RelationSmoother(
    private val emaWeight: Float = 0.62f,
    private val maxMisses: Int = 1,
    private val maxRelations: Int = 3,
) {
    private data class Entry(
        var relation: SceneRelation,
        var score: Float,
        var rankScore: Float,
        var misses: Int,
    )

    private val entries = linkedMapOf<String, Entry>()

    fun reset() = entries.clear()

    fun update(relations: List<SceneRelation>): List<SceneRelation> {
        val seen = mutableSetOf<String>()

        relations.forEach { relation ->
            val subjectId = relation.subject.trackId ?: return@forEach
            val objectId = relation.objectDetection.trackId ?: return@forEach
            val key = "$subjectId|${relation.predicate}|$objectId"
            seen += key

            val previous = entries[key]
            if (previous == null) {
                entries[key] = Entry(relation, relation.score, relation.rankScore, 0)
            } else {
                previous.score = previous.score * emaWeight + relation.score * (1f - emaWeight)
                previous.rankScore = previous.rankScore * emaWeight + relation.rankScore * (1f - emaWeight)
                previous.relation = relation.copy(score = previous.score, rankScore = previous.rankScore)
                previous.misses = 0
            }
        }

        entries.entries.toList().forEach { (key, entry) ->
            if (key !in seen) {
                entry.misses += 1
                if (entry.misses > maxMisses) entries.remove(key)
            }
        }

        return entries.values
            .sortedByDescending { it.rankScore }
            .take(maxRelations)
            .map { it.relation.copy(score = it.score, rankScore = it.rankScore) }
    }

    fun resolveBoxes(relations: List<SceneRelation>, detections: List<Detection>): List<SceneRelation> {
        val byId = detections.mapNotNull { detection ->
            detection.trackId?.let { it to detection }
        }.toMap()

        return relations.mapNotNull { relation ->
            val subjectId = relation.subject.trackId ?: return@mapNotNull relation
            val objectId = relation.objectDetection.trackId ?: return@mapNotNull relation
            val subject = byId[subjectId] ?: return@mapNotNull null
            val objectDetection = byId[objectId] ?: return@mapNotNull null
            relation.copy(subject = subject, objectDetection = objectDetection)
        }
    }
}
